react-app/src/library/services/headset.ts

```diff
@@ -0,0 +1,203 @@
+    implementation.on('webHidPermissionRequested' as any, (payload: any) => {
+      this.logger.debug('Requesting Webhid Permissions');
+      this._headsetEvents$.next({ event: 'webHidPermissionRequested' as any, payload })
+    });
+  }
+
+  activeMicChange(newMicLabel: string): void {
+    const implementation = this.buildImplementationsArray().find((implementation) => implementation.deviceLabelMatchesVendor(newMicLabel));
+    if (implementation) {
+      this.changeImplementation(implementation, newMicLabel);
+    } else if (this.selectedImplementation) {
+      this.selectedImplementation.disconnect();
+    }
+  }
+
+  private performActionIfConnected(actionName: string, perform: (impl: VendorImplementation) => Promise<any>) {
+    const impl = this.selectedImplementation;
+    if(!impl || !impl.isConnected) {
+      this.logger.info(`Headset: No vendor headset connected [${actionName}]`);
+      return Promise.resolve();
+    }
+
+    return perform(impl);
+  }
+
+  async changeImplementation(implementation: VendorImplementation | null, deviceLabel: string): Promise<void> {
+    if (implementation === this.selectedImplementation) {
+      return;
+    }
+
+    if (this.selectedImplementation) {
+      this.selectedImplementation.disconnect();
+    }
+
+    this.selectedImplementation = implementation;
+
+    if (implementation) {
+      implementation.connect(deviceLabel);
+    }
+
+    this._headsetEvents$.next({ event: 'implementationChanged', payload: implementation});
+  }
+
+  incomingCall(callInfo: CallInfo, hasOtherActiveCalls?: boolean): Promise<any> {
+    return this.performActionIfConnected('incomingCall', (implementation) => implementation.incomingCall(callInfo, hasOtherActiveCalls));
+  }
+
+  outgoingCall(callInfo: CallInfo): Promise<any> {
+    return this.performActionIfConnected('outgoingCall', (implementation) => implementation.outgoingCall(callInfo));
+  }
+
+  answerCall(conversationId: string): Promise<any> {
+    return this.performActionIfConnected('answerCall', (implementation) => implementation.answerCall(conversationId));
+  }
+
+  setMute(value: boolean): Promise<any> {
+    return this.performActionIfConnected('setMute', (implementation) => implementation.setMute(value));
+  }
+
+  setHold(conversationId: string, value: boolean): Promise<any> {
+    return this.performActionIfConnected('setHold', (implementation) => implementation.setHold(conversationId, value));
+  }
+
+  endCall(conversationId: string, hasOtherActiveCalls?: boolean): Promise<any> {
+    return this.performActionIfConnected('endCall', (implementation) => implementation.endCall(conversationId, hasOtherActiveCalls));
+  }
+
+  endAllCalls(): Promise<any> {
+    return this.performActionIfConnected('endAllCalls', (implementation) => implementation.endAllCalls());
+  }
+
+  private handleDeviceAnsweredCall(event: VendorEvent<EventInfo>): void {
+    console.log('event answered Call -> ', event);
```

Comment:
NIT: Can this use the logger instead?

Reply:
This was another test log I put in and totally forgot to remove 😅